test/server_test.py

```diff
@@ -106,12 +112,41 @@ def missingCSRFToken():
 			})
 			assertResponse(response, 400, 'Session expired. Reload and try again')
 
+		@it('Empty body')
+		def emptyBody():
+			res1 = app.post('/login/form')
+			res2 = app.post('/login/form', data={})
+			res3 = app.post('/login/form', json={})
+			assertResponse(res1, 400, 'Session expired. Reload and try again')
+			assertResponse(res2, 400, 'Session expired. Reload and try again')
+			assertResponse(res3, 400, 'Session expired. Reload and try again')
+
+		@it('Non-json body')
+		def incorrectData():
+			res = app.post('/login/form', data='I am not json')
+			assertResponse(res, 400, 'Session expired. Reload and try again')
+
+		@it('Hitting rate limit')
+		def rateLimit():
+			enableRateLimiter(True)
+			data = {
+				'csrf_token': getLoginCSRFToken(),
+				'username': test_name,
+				'password': test_pass
+			}
+			for i in range(0, 10):
```

Comment:
I feel this range should be gathered from reading the config file. Because when that part of the config changes this test will become invalid.
If not read from the config, we should at least use a global constant to unify all of these for manageability.

Sidenote: If your range is 0 to X, you can omit the first value: range(0, 10) -> range(10).